add canary release cookbook (#959)
megaease/easegress: doc/cookbook/canary-release.md

    servers:
    - url: http://megaease.com/api/v1 # old version
- name: buildResponse
  kind: ResponseBuilder
  template: |
    statusCode: {{.responses.order.StatusCode}}
    body: |
      [{{.responses.notify.Body}}]
```

Then create the E-commerce pipeline with the command:

```bash
egctl object create -f ecommerce-service.yaml
```

Save below YAML to ecommerce-server.yaml.

```yaml
kind: HTTPServer
name: ecommerce-server
port: 8080
https: false
keepAlive: true
keepAliveTimeout: 75s
maxConnection: 10240
cacheSize: 0
rules:
  - paths:
    - pathPrefix: /ecommerce
      backend: ecommerce-pipeline
```

Then create the HTTP server with command:

```bash
egctl object create -f ecommerce-server.yaml
```

## based on user devices

![case 2](../imgs/canary-release-case-2.png)

In the real world, services are rarely ever standalone entities, but rather are often interconnected and dependent on one another to provide a holistic experience for the customer. As shown, we have recently upgraded our notify service from v1 to v2. we would like to direct users who are accessing our service from a Mac device to use the new notify service (v2).

> 1. Traffic tagging: the user device judgment is based on the `user-agent` field in HTTP request. We can get user device and system from the user-agent filed of HTTP request by [Workers](https://developers.cloudflare.com/workers/).

Reviewer: ```suggestion
**1. Traffic tagging**

To determine the user's device and system, we can extract it from the `User-Agent` field in HTTP request by [Workers](https://developers.cloudflare.com/workers/).
```

> We provided easy way to parse user-agent in [easegress-rust-uaparser](https://github.com/megaease/easegress-rust-uaparser). Download `easegress-rust-uaparser`

Reviewer: ```suggestion
We offer an easy way to parse `User-Agent` in [easegress-rust-uaparser](https://github.com/megaease/easegress-rust-uaparser) library. 
```

```bash
git clone https://github.com/megaease/easegress-rust-uaparser.git
```

> Enter the `easegress-rust-uaparser/cloudflare` directory and deploy the worker using the wrangler command

Reviewer: ```suggestion
Go to the `easegress-rust-uaparser/cloudflare` directory and deploy the worker by using the `wrangler` command
```

```bash
npx wrangler publish
```

> Enter the Cloudflare website configuration page and configure Workers Routes as you needed.

Reviewer: ```suggestion
Go to the Cloudflare website configuration page and configure _Workers Routes_.
```

> 2. Traffic scheduling rules:

Reviewer: ```suggestion
**2. Traffic scheduling rules**
```

> Save the below YAML to ecommerce-service.yaml, and make sure you have replaced the servers with yours.

Reviewer: ```suggestion
Save the following YAML code to `ecommerce-service.yaml`, and make sure to replace the `servers` configuration with your own.
```

  namespace: order
- filter: buildNotifyRequest
  namespace: notify
- filter: notify-service
  namespace: notify
- filter: buildResponse
filters:
- name: buildOrderRequest
  kind: RequestBuilder
  template: |
    url: /order
- name: order-service
  kind: Proxy
  pools:
  - servers:
    - url: http://megaease.com/api/v2
- name: buildNotifyRequest
  kind: RequestBuilder  # use order response as notify request body
  template: |
    url: /notify
    method: POST
    body: |
      {"order": "{{.responses.order.Body | jsonEscape}}"}
- name: notify-service
  kind: Proxy
  pools:
  - servers:
    - url: http://megaease.com/api/v1
  - filter:
      headers:
        x-ua-device: # Cloudflare's user device header
          exact: Mac # the traffic from Mac should be directed to the new version
    servers:
    - url: http://megaease.com/api/v2
- name: buildResponse
  kind: ResponseBuilder
  template: |
    statusCode: {{.responses.notify.StatusCode}}
    body: |
      [{{.responses.notify.Body}}]
```

Then update the E-commerce pipeline with the command:

```bash
egctl object update -f ecommerce-service.yaml
```

## based on user OS

![case 3](../imgs/canary-release-case-3.png)

> Suppose order service and notify service have upgraded at the same time. The user who is accessing our service from a Mac device should be directed to the new version of the order service (v2), and from MacOS should be directed to the new version of the notify service (v2). 

Reviewer: ```suggestion
Suppose both the Order service and Notify service have been upgraded simultaneously. The user who is accessing our service from a Mac device should be routed to the new version of the Order service (v2), and from MacOS should be routed to the new version of the Notify service (v2). 
```

> You may think Cloudflare's Worker is not suitable for you, there is another easy way to parse user-agent with Easegress's WASM.

Reviewer: ```suggestion
You may think Cloudflare's Workers is not suitable for your needs, there is another easy way to parse `User-Agent` with Easegress's WASM.
```

> 1. Traffic tagging: using Easegress's WASM to parse user-agent. 

Reviewer: ```suggestion
**1. Traffic tagging** 

Using Easegress's WASM to parse the `User-Agent`. 
```

Download `easegress-rust-uaparser`

```bash
git clone https://github.com/megaease/easegress-rust-uaparser.git
```

> Navigate to the `easegress-rust-uaparser/binary` directory and use the `easegress.wasm` file directly. It will parse user-agent and add the `x-ua-device` and `x-ua-os` headers to the HTTP request.

Reviewer: ```suggestion
Go to the `easegress-rust-uaparser/binary` directory and just use the `easegress.wasm` file directly. It will parse the `User-Agent` and add the `x-ua-device` and `x-ua-os` headers to the HTTP request.
```

> 2. Traffic scheduling rules:

Reviewer: ```suggestion
**2. Traffic scheduling rules**
```

> Save the below YAML to ecommerce-service.yaml, and make sure you have replaced the servers and wasm file path with yours.

Reviewer: ```suggestion
Save the following YAML code to `ecommerce-service.yaml`, and make sure to replace the `servers` configuration and wasm file path with your own.
```

* **Traffic scheduling**: Once traffic is tagged, we can use traffic rules to schedule specific traffic to a designated Canary. For instance, traffic from an iPhone user in Beijing could be directed to a new version of a service.

## Why use Easegress and Cloudflare

* For traffic tagging, Using [Cloudflare](https://www.cloudflare.com/) we can easily label the traffic.
  - **Geographic Location**: Cloudflare offers a feature called [Managed Transforms](https://developers.cloudflare.com/rules/transform/managed-transforms/reference/), which can be used to identify the geographic location of a user.
  - **User Device**: Using [Workers](https://developers.cloudflare.com/workers/), we can extract information about the user's device, operating system, and web browser from the `User-Agent` field of the HTTP request.
  - **User's Customized Label**: Using Cloudflare [Rules](https://developers.cloudflare.com/rules/transform/request-header-modification/), it is easy to extract the user's customized business label from HTTP request headers (such as: Cookies, etc.)

> - For traffic scheduling, Easegress offers two features, `Filter` and `Proxy`, that can be used to route traffic to different canary releases.

Reviewer 2: This line is not accurate, the HTTP Server can also be used for traffic routing. And for `Filter`, do you mean the `filter` property of a server pool in a proxy or a filter in the pipeline? So I propose not providing too many details here.

```suggestion
- For traffic scheduling, Easegress can be used to route traffic to different canary releases.
```
